lib/CPathfinder.cpp

```diff
@@ -79,96 +86,118 @@ void CPathfinder::calculatePaths()
 
-			int turnAtNextTile = turn;
-			if(remains < 0)
-			{
-				//occurs rarely, when hero with low movepoints tries to leave the road
-				turnAtNextTile++;
-				int moveAtNextTile = maxMovePoints(cp);
-				cost = gs->getMovementCost(hero, cp->coord, dp->coord, moveAtNextTile); //cost must be updated, movement points changed :(
-				remains = moveAtNextTile - cost;
-			}
+				if(!hlp->isLayerAvailable(i))
+					continue;
 
-			if(isBetterWay(remains, turnAtNextTile))
-			{
-				assert(dp != cp->theNodeBefore); //two tiles can't point to each other
-				dp->moveRemains = remains;
-				dp->turns = turnAtNextTile;
-				dp->theNodeBefore = cp;
+				if(cp->layer != i && !isLayerTransitionPossible())
+					continue;
+
+				if(!isMovementToDestPossible())
+					continue;
+
+				destAction = getDestAction();
+				int cost = CPathfinderHelper::getMovementCost(hero, cp->coord, dp->coord, ct, dt, movement, hlp->getTurnInfo());
+				int remains = movement - cost;
+				if(destAction == CGPathNode::EMBARK || destAction == CGPathNode::DISEMBARK)
+				{
+					remains = hero->movementPointsAfterEmbark(movement, cost, destAction - 1, hlp->getTurnInfo());
+					cost = movement - remains;
+				}
+				int turnAtNextTile = turn;
+				if(remains < 0)
+				{
+					//occurs rarely, when hero with low movepoints tries to leave the road
+					hlp->updateTurnInfo(++turnAtNextTile);
+					int moveAtNextTile = hlp->getMaxMovePoints(i);
+					cost = CPathfinderHelper::getMovementCost(hero, cp->coord, dp->coord, ct, dt, moveAtNextTile, hlp->getTurnInfo()); //cost must be updated, movement points changed :(
+					remains = moveAtNextTile - cost;
+				}
+
+				if(isBetterWay(remains, turnAtNextTile)
+					&& passOneTurnLimitCheck(cp->turns != turnAtNextTile || !remains))
+				{
+					assert(dp != cp->theNodeBefore); //two tiles can't point to each other
+					dp->moveRemains = remains;
+					dp->turns = turnAtNextTile;
+					dp->theNodeBefore = cp;
+					dp->action = destAction;
 
-				if(checkDestinationTile())
-					mq.push_back(dp);
+					if(isMovementAfterDestPossible())
+						pq.push(dp);
+				}
 			}
 		} //neighbours loop
 
 		//just add all passable teleport exits
-		if(sTileObj)
+		addTeleportExits();
+		for(auto & neighbour : neighbours)
 		{
-			addTeleportExits();
-			for(auto & neighbour : neighbours)
+			dp = out.getNode(neighbour, cp->layer);
+			if(dp->locked)
+				continue;
+
+			if(isBetterWay(movement, turn))
 			{
-				dp = getNode(neighbour);
-				if(isBetterWay(movement, turn))
-				{
-					dp->moveRemains = movement;
-					dp->turns = turn;
-					dp->theNodeBefore = cp;
-					mq.push_back(dp);
-				}
+				dp->moveRemains = movement;
+				dp->turns = turn;
+				dp->theNodeBefore = cp;
+				dp->action = CGPathNode::NORMAL;
+				pq.push(dp);
 			}
 		}
 	} //queue loop
 }
 
-void CPathfinder::addNeighbours(const int3 &coord)
+void CPathfinder::addNeighbours()
 {
 	neighbours.clear();
-	ct = &gs->map->getTile(coord);
-
 	std::vector<int3> tiles;
```

Comment:
You reuse "neightbours" but reallocate tiles vector every time, which is time-consuming operation.